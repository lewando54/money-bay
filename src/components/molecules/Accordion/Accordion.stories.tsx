/* eslint-disable @typescript-eslint/consistent-type-assertions */
import { StoryObj, Meta } from '@storybook/react-native'

import Accordion from './Accordion'

const meta = {
	component: Accordion,
	argTypes: {
        // To write
	}
} satisfies Meta<typeof Accordion>

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
	args: {
        // To write
	}
}
