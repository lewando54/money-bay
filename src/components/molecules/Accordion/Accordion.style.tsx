import { StyleSheet } from 'react-native'
export default StyleSheet.create({
    mainContainer: {
        backgroundColor: 'white',
        borderRadius: 10,
        overflow: 'hidden'
    },
    headerContainer: {
        padding: 20,
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'center'
    },
    bodyText: {
        position: 'absolute',
        width: '100%',
        top: 0,
        padding: 20
    }
})
