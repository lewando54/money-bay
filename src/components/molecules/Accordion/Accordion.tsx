import React from 'react'
import { useState } from 'react'
import { Pressable,  View } from 'react-native'
import Animated, { interpolate, interpolateColor, measure, runOnUI, useAnimatedRef, useAnimatedStyle, useSharedValue, withTiming } from 'react-native-reanimated'
import { Entypo } from '@expo/vector-icons'

import { CustomText } from '@atoms'

import accordionStyle from './Accordion.style'

import globalThemeStyle, { globalColors } from '@styling/GlobalTheme.style'

interface IAccordionProps {
    extended?: boolean
    headerText?: string
    bodyText?: string
}

export default function Accordion({
    extended=false,
    headerText='How does the identification code process work?',
    bodyText='Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\n\nDuis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatu dolor in repreh enderitr lla cca ecat cupidatat non proiden paria.'
}: IAccordionProps){
    const [isExtended, setIsExtended] = useState(extended)
    const textRef = useAnimatedRef<Animated.View>()
    const heightValue = useSharedValue(0)
    const rotationValue = useSharedValue(0)

    const heightAnimationStyle = useAnimatedStyle(() => ({
        height: heightValue.value,
    }))

    const textColorAnimationStyle = useAnimatedStyle(() => ({
        color: interpolateColor(rotationValue.value, [0, 1], ['black', 'red'])
    }))

    const separatorAnimationStyle = useAnimatedStyle(() => ({
        borderTopWidth: rotationValue.value
    }))

    const rotationAnimationStyle = useAnimatedStyle(() => ({
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        transform: [{rotateZ: interpolate(rotationValue.value, [0, 1], [0, 180]) + 'deg'}]
    }))

    const onPressHandle = () => {
        if (!isExtended) {
            runOnUI(() => {
                'worklet'
                if(measure(textRef)){
                    heightValue.value = withTiming(measure(textRef)!.height)
                } 
                else {
                    heightValue.value = withTiming(100)
                }
            })()
            rotationValue.value = withTiming(1)
        } else {
            heightValue.value = withTiming(0)
            rotationValue.value = withTiming(0)
        }
        setIsExtended(() => !isExtended)
    }

    return (
        <Pressable onPress={onPressHandle}>
            <Animated.View style={accordionStyle.mainContainer}>
                <View style={accordionStyle.headerContainer} testID='acc-header'>
                    <Animated.Text style={[globalThemeStyle.text_Title, textColorAnimationStyle]}>{headerText}</Animated.Text>
                    <Animated.View style={rotationAnimationStyle}>
                        <Entypo name="chevron-small-down" size={24}/>
                    </Animated.View>
                </View>
                <Animated.View testID='acc-body' style={[{borderTopColor: '#CED6E1'}, heightAnimationStyle, separatorAnimationStyle]}>
                    <Animated.View ref={textRef} style={accordionStyle.bodyText}>
                        <CustomText style={[globalThemeStyle.text_Regular, {color: globalColors.BODY_TEXT_COLOR}]}>{bodyText}</CustomText>
                    </Animated.View>
                </Animated.View>
            </Animated.View>
        </Pressable>
    )
}
