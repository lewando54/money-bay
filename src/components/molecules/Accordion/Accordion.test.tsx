import React from 'react'
import renderer from 'react-test-renderer'

import Accordion from './Accordion'
describe('Accordion', () => {
    it('should render correctly', () => {
        const tree = renderer.create(<Accordion />)
        expect(tree).toMatchSnapshot()
    })
})
