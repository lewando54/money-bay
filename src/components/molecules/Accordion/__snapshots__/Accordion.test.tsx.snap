// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    animatedStyle={
      {
        "value": {},
      }
    }
    collapsable={false}
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 10,
        "overflow": "hidden",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 20,
        }
      }
      testID="acc-header"
    >
      <Text
        animatedStyle={
          {
            "value": {
              "color": "rgba(0, 0, 0, 1)",
            },
          }
        }
        collapsable={false}
        style={
          {
            "color": "rgba(0, 0, 0, 1)",
            "fontFamily": "MulishSemiBold",
            "fontSize": 14,
          }
        }
      >
        How does the identification code process work?
      </Text>
      <View
        animatedStyle={
          {
            "value": {
              "transform": [
                {
                  "rotateZ": "0deg",
                },
              ],
            },
          }
        }
        collapsable={false}
        style={
          {
            "transform": [
              {
                "rotateZ": "0deg",
              },
            ],
          }
        }
      >
        <Text />
      </View>
    </View>
    <View
      animatedStyle={
        {
          "value": {
            "borderTopWidth": 0,
            "height": 0,
          },
        }
      }
      collapsable={false}
      style={
        {
          "borderTopColor": "#CED6E1",
          "borderTopWidth": 0,
          "height": 0,
        }
      }
      testID="acc-body"
    >
      <View
        animatedStyle={
          {
            "value": {},
          }
        }
        collapsable={false}
        forwardedRef={[Function]}
        style={
          {
            "padding": 20,
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            [
              {
                "color": "#1B1D4D",
                "fontFamily": "MulishRegular",
                "fontSize": 16,
              },
              [
                {
                  "color": "white",
                  "fontFamily": "MulishRegular",
                },
                {
                  "color": "#4C4C60",
                },
              ],
            ]
          }
        >
          Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatu dolor in repreh enderitr lla cca ecat cupidatat non proiden paria.
        </Text>
      </View>
    </View>
  </View>
</View>
`;
